### DISCUSSION TITLE
Why am I using sqlalchemy to query clickhouse so much slower than clickhouse_drive

### DISCUSSION THREAD

**Discussion by 945226956** [sqlalchemy/sqlalchemy#8937] [Usage Questions]
I have an sql that has about 103800 rows, 13 columns and 100M. When I query pandas.read_sql(sql, conn) and sqlalchemy.create_engine.execute(sql), it takes 65 seconds. But when I query with clickhouse_driver.Client.query_dataframe(sql) it only takes 9s. Why is the gap so wide? In addition, I used another tool (dbvear) to perform sql changes that never took nearly a minute.

sql:

SELECT
	a.task_id as task_id,
	a.channel as channel,
	a.return_n as return_n,
	a.time_arr as time_arr,
	a.breath_data_rt_v4 as breath_data_rt_v4,
	a.breath_data_rt_v5 as breath_data_rt_v5,
	case
		when b.tw is null then 0.15
		else toFloat64(b.tw)
	end as tw,
	b.interval as interval,
	b.start_time_str as start_time_str,
	c.discard_bytes_index / 4 / 48000 as discard_second,
	c.discard_bytes_index / 4 / 48000 + a.time_arr as data_second,
	DATE_ADD( SECOND, toFloat64(c.discard_bytes_index / 4 / 48000), b.start_time_str) AS wav_start_time,
	DATE_ADD( SECOND,(toFloat64((c.discard_bytes_index / 4 / 48000)+ a.time_arr)), b.start_time_str) AS time_arr_str, 
	1 as t
FROM
	log_df a
left join session_info_base b on
	a.task_id = b.task_id
left join session_info_exit c on
	a.task_id = c.task_id
WHERE
	a.task_id = '2022-12-05 23:40:45@403414'
ORDER BY
	channel,
	return_n,
	time_arr;

SOME CODE:
<img width="740" alt="image" src="https://user-images.githubusercontent.com/21196092/205873019-bb2d1ab8-4679-498c-9570-70f437b0d543.png">
<img width="702" alt="image" src="https://user-images.githubusercontent.com/21196092/205873257-8848d121-9f16-4728-92d4-625b8d0cc085.png">

<img width="1004" alt="image" src="https://user-images.githubusercontent.com/21196092/206193486-214a9cf4-1342-48e4-ba14-e4588d2e4842.png">

**Comment by zzzeek** [comment]
I wouldnt know on this end, it looks like you are actually comparing the implementations of the functions `clickhouse_driver.Client.query_dataframe()` vs. the Pandas function `pandas.read_sql()`.   It likely has a more to do with those two functions that with SQLAlchemy, which using Core can select rows and fetch data from them not quite as fast as the raw driver, but not by an order of magnitude like that.


for SQLAlchemy's side of it work with the [performance examples](https://docs.sqlalchemy.org/en/14/orm/examples.html#module-examples.performance) to get a feel for the relative speed of the different approaches.  for a guide to profiling see https://docs.sqlalchemy.org/en/14/faq/performance.html#how-can-i-profile-a-sqlalchemy-powered-application

**Reply by 945226956** [reply]
good page,I will try it !
